### ISSUE TITLE
[RFC]: CodePredictorAttention Abstraction

### ISSUE THREAD

**Issue by gcanlin** [vllm-project/vllm-omni#2967]
# RFC: CodePredictorAttention Abstraction

Part of #2287.

---

## 1. Summary

Introduce a dedicated abstraction for `CodePredictorAttention` that keeps the code predictor model pure while allowing platform-specific attention implementations:

- `SDPA` on GPU/CPU
- `npu_fusion_attention` on Ascend NPU

This RFC does **not** propose reusing the full vLLM decoder attention stack. It instead proposes a lighter-weight abstraction closer in spirit to `MMEncoderAttention`: a standalone attention component with pluggable backends and minimal runtime-specific surface area.

---

### 2.1 Current Situation

`CodePredictorAttention` in `vllm_omni/model_executor/models/common/qwen3_code_predictor.py` is a model-private self-attention implementation used by the TTS code predictor. It has the following characteristics:

- causal self-attention
- no KV cache
- short sequence length
- full re-prefill on every AR step
- no dependency on the vLLM paged attention scheduler

This is fundamentally different from the main LLM attention path in vLLM.

### 2.2 Why the Main vLLM Attention Stack Is Not a Good Fit

The standard vLLM decoder attention stack is built around:

- per-layer attention backends
- KV cache management
- `attn_metadata`
- block tables and slot mapping
- model-runner-managed decode/prefill scheduling

`CodePredictorAttention` does not need that machinery. Forcing it into the full decoder attention abstraction would add complexity without adding real value.

### 2.3 Why the Current Inline Implementation Is Also Not Good Enough

The current direct SDPA call is simple, but it does not provide a clean place for:

- NPU-specific operator selection
- GQA expansion when backend support differs
- causal mask strategy differences
- operator-specific layout conversion
- future backend-specific optimizations

If NPU support is added directly inside `CodePredictorAttention.forward()` using `if is_npu`, the model layer becomes platform-aware again, which is the opposite of the intended architecture direction.

---

### 3.1 Primary Goal

Make `CodePredictorAttention` platform-agnostic while preserving the current compute semantics.

### 3.2 Non-Goals

This RFC does **not** aim to:

- reuse the full vLLM decoder `AttentionBackend` + `AttentionMetadataBuilder` stack
- introduce KV cache support for the code predictor
- make the code predictor part of the multimodal encoder cache pipeline
- solve graph abstraction in this document beyond defining clean boundaries

---

### 4.1 Closer to `MMEncoderAttention` Than to Decoder Attention

The closest precedent in vLLM is `MMEncoderAttention`, not decoder attention.

What `CodePredictorAttention` shares with `MMEncoderAttention`:

- standalone attention module inside a submodel
- no KV cache
- whole-sequence forward
- strong platform dependence at the operator level
- need for lightweight backend selection

What it does **not** share with `MMEncoderAttention`:

- vision-style packed variable-length batching
- non-causal attention semantics
- `cu_seqlens`-driven encoder batch organization

So the right conclusion is:

- borrow the **component-level abstraction style** from `MMEncoderAttention`
- do **not** copy its exact input contract

### 4.2 Keep Platform Logic Out of the Model

The model should say only:

> compute causal self-attention for these `q/k/v` tensors

It should not say:

- which backend to use
- how to expand GQA for a specific backend
- whether a fixed causal mask is required
- what tensor layout a device kernel expects

### 4.3 Keep Graph Concerns Separate

This RFC concerns only the attention operator abstraction.

Graph capture and replay should remain outside the attention component. The attention backend may expose constraints that are relevant to graph capture, but the backend should not own graph orchestration.

---

### 5.1 High-Level Structure

```text
CodePredictorAttention
    |
    +-- CodePredictorAttnBackend
            |
            +-- SDPACodePredictorBackend
            +-- NPUFusionCodePredictorBackend
```

### 5.2 Responsibilities

`CodePredictorAttention`

- owns projections, RoPE application, and output projection
- reshapes tensors into a canonical logical form
- delegates backend-specific attention execution

`CodePredictorAttnBackend`

- owns backend-specific operator invocation
- owns backend-specific GQA handling if needed
- owns backend-specific layout conversion
- owns backend-specific causal masking strategy

---

## 6. Canonical Attention Contract

The canonical input contract for the backend should be:

- `query`: `[B, num_heads, S, D]`
- `key`: `[B, num_kv_heads, S, D]`
- `value`: `[B, num_kv_heads, S, D]`

The backend returns:

- `output`: `[B, num_heads, S, D]`

This contract is intentionally simple and stable. It matches the natural internal representation already used in the code predictor after projection and RoPE.

This also means:

- the backend does not own projection layers
- the backend does not own RoPE
- the backend does not own output projection

---

## 7. Backend Interface

```python
from abc import ABC, abstractmethod
from dataclasses import dataclass
import torch


@dataclass
class CodePredictorAttnConfig:
    num_heads: int
    num_kv_heads: int
    head_dim: int
    scale: float
    max_seq_len: int


class CodePredictorAttnBackend(ABC):
    @staticmethod
    @abstractmethod
    def get_name() -> str:
        raise NotImplementedError

    @classmethod
    @abstractmethod
    def is_available(cls) -> bool:
        raise NotImplementedError

    @abstractmethod
    def forward(
        self,
        query: torch.Tensor,
        key: torch.Tensor,
        value: torch.Tensor,
    ) -> torch.Tensor:
        raise NotImplementedError
```

---

### 8.1 `SDPACodePredictorBackend`

This is the default backend for GPU/CPU.

Expected behavior:

- call `torch.nn.functional.scaled_dot_product_attention`
- use `is_causal=True`
- use `enable_gqa=True` only when supported and intended
- preserve current numerical behavior

Pseudo-implementation:

```python
class SDPACodePredictorBackend(CodePredictorAttnBackend):
    def __init__(self, config: CodePredictorAttnConfig):
        self.config = config
        self.use_gqa = config.num_heads != config.num_kv_heads

    def forward(self, query, key, value):
        return F.scaled_dot_product_attention(
            query,
            key,
            value,
            scale=self.config.scale,
            is_causal=True,
            enable_gqa=self.use_gqa,
        )
```

### 8.2 `NPUFusionCodePredictorBackend`

This is the Ascend backend.

Expected behavior:

- use `torch_npu.npu_fusion_attention`
- perform explicit GQA expansion if required
- perform backend-required layout conversion
- use an explicit causal masking strategy compatible with the chosen NPU operator path

Important note:

This RFC intentionally does **not** mandate a packed `TND + actual_seq_lengths` interface like `AscendMMEncoderAttention`.

That pattern is a good fit for variable-length encoder batches, but `CodePredictorAttention` is a short, causal, re-prefill self-attention path. A simpler causal self-attention invocation is preferred.

Pseudo-implementation:

```python
class NPUFusionCodePredictorBackend(CodePredictorAttnBackend):
    def __init__(self, config: CodePredictorAttnConfig):
        self.config = config
        self.use_gqa = config.num_heads != config.num_kv_heads
        self.repeat = config.num_heads // config.num_kv_heads
        self.causal_mask = ...

    def forward(self, query, key, value):
        if self.use_gqa:
            key = key.repeat_interleave(self.repeat, dim=1)
            value = value.repeat_interleave(self.repeat, dim=1)

        out = torch_npu.npu_fusion_attention(
            query=query,
            key=key,
            value=value,
            head_num=self.config.num_heads,
            input_layout="BNSD",
            atten_mask=self.causal_mask,
            scale=self.config.scale,
            ...
        )[0]
        return out
```

---

## 9. Why Not Reuse `AscendMMEncoderAttention` Directly

`AscendMMEncoderAttention` is useful as a reference, but it should not be reused directly for the code predictor.

Reasons:

1. It is optimized for encoder-style attention, not causal self-attention.
2. It is built around packed variable-length inputs.
3. Its `TND` layout and `actual_seq_qlen` / `actual_seq_kvlen` interface primarily serve multimodal encoder batching.
4. The code predictor has a simpler and more stable input pattern.

What should be reused conceptually:

- `npu_fusion_attention` as the NPU fast path
- explicit GQA expansion
- optional head-dim padding if needed by kernel constraints
- component-level replacement instead of model-level platform branching

What should not be copied mechanically:

- the exact `MMEncoderAttention` input metadata contract
- the multimodal encoder batching assumptions

---

## 10. Backend Selection

Backend selection should be local and simple.

Pseudo-API:

```python
def get_code_predictor_attn_backend(
    config: CodePredictorAttnConfig,
) -> CodePredictorAttnBackend:
    if current_omni_platform.is_npu():
        from ...npu_fusion import NPUFusionCodePredictorBackend
        if NPUFusionCodePredictorBackend.is_available():
            return NPUFusionCodePredictorBackend(config)

    return SDPACodePredictorBackend(config)
```

This keeps the selection logic:

- explicit
- lightweight
- independent from the main decoder attention selector

---

## 11. Integration into `CodePredictorAttention`

`CodePredictorAttention` should be simplified to:

1. project hidden states to `q/k/v`
2. apply normalization
3. apply RoPE
4. call `self.attn_backend.forward(q, k, v)`
5. reshape output and apply `o_proj`

Pseudo-structure:

```python
class CodePredictorAttention(nn.Module):
    def __init__(self, config, ...):
        ...
        self.attn_backend = get_code_predictor_attn_backend(
            CodePredictorAttnConfig(
                num_heads=self.num_heads,
                num_kv_heads=self.num_kv_heads,
                head_dim=self.head_dim,
                scale=self.scaling,
                max_seq_len=config.num_code_groups + 1,
            )
        )

    def forward(self, hidden_states, position_embeddings):
        ...
        attn_out = self.attn_backend.forward(q, k, v)
        ...
```

This is the central architectural outcome of this RFC.

---

## 12. Relationship to Graph Abstraction

This RFC is intentionally narrower than the TTS graph abstraction RFC.

The interaction boundary is:

- this RFC defines how `CodePredictorAttention` becomes platform-agnostic
- graph RFCs define how the code predictor model or executor is captured/replayed

The two proposals are complementary:

- attention backend abstraction removes operator branching from the model
- runner-side graph abstraction removes graph management from the model

Together they produce a clean separation of concerns.

---

### Phase 1: Introduce Attention Backend Interface

Add:

- `code_predictor.py`
- `sdpa.py`
- `selector.py`
- NPU backend implementation in the Ascend side

### Phase 2: Refactor `CodePredictorAttention`

Replace direct SDPA calls with backend delegation.

### Phase 3: Remove Model-Level Platform Branching

Ensure there is no `if is_npu` logic left inside the model attention path.

### Phase 4: Align with Runner-Side Graph Abstraction

Once the attention component is backendized, graph abstraction can evolve independently at the runner layer.

---

## 14. Benefits

1. `CodePredictorAttention` remains model-centric and compute-only.
2. NPU fast paths can be added without polluting the model definition.
3. The abstraction matches the real problem shape better than the main decoder attention stack.
4. The design is small enough to be practical, but structured enough to scale.
5. It aligns naturally with future runner-side graph cleanup.

---

## 15. Open Questions

1. Should NPU use `BNSD` or another layout for the final implementation?
2. Should head-dim padding be part of the backend now, or added only if required by kernel constraints?
3. Should the backend expose capability flags relevant to graph capture later, or should that stay entirely outside the attention layer?
4. Should this abstraction remain code-predictor-specific, or later be generalized into a small reusable "causal no-cache attention op" utility?

---

## 16. Recommendation

Adopt a lightweight `CodePredictorAttention` backend abstraction modeled after the component-level philosophy of `MMEncoderAttention`, but tailored for causal short-sequence self-attention.

In short:

- do **not** force the code predictor into the full vLLM decoder attention stack
- do **not** keep platform-specific operator branching inside the model
- do build a small, explicit backend abstraction for `CodePredictorAttention`

**Comment by gcanlin** [comment]
cc @hsliuustc0106 @ZeldaHuang @linyueqian @Sy0307 @Isotr0py

**Comment by Sy0307** [comment]
LGTM. Different backend is good idea. Some optimization on GPU may not run on NPU. In future model adaption roadmap, we should keep two path: Basic path suitable for different hardware and Optimized path for GPU or NPU or XPU.
